rp2040-hal/src/spi.rs

```diff
@@ -1,3 +1,184 @@
+    fn init(&mut self, resets: &mut RESETS, baudrate: u32);
+
+    ///
+    fn set_baudrate(&mut self, baudrate: u32) -> u32;
+
+    ///
+    fn set_format(&mut self, data_bits: u8, cpol: Cpol, cpha: Cpha);
+
+    ///
+    fn is_writable(&self) -> bool;
+
+    ///
+    fn is_readable(&self) -> bool;
+
+    ///
+    fn is_busy(&self) -> bool;
+}
+
+///
+pub enum Cpol {
+    /// CPOL=0
+    _0 = 0,
+    /// CPOL=1
+    _1,
+}
+
+///
+pub enum Cpha {
+    ///CPHA=0
+    _0 = 0,
+    ///CPHA=1
+    _1,
+}
+
+macro_rules! impl_spi {
+    ($name:ident, $dev:ident, $description:literal) => {
+        /// Spi 0
+        #[doc()]
+        pub struct $name {
+            spi_dev: $dev,
+        }
+
+        impl $name {
+            /// Create new spi device
+            pub fn new(spi_dev: $dev) -> Self {
+                $name { spi_dev }
+            }
+
+            /// Free underlying spi registers
+            pub fn free(self) -> $dev {
+                self.spi_dev
+            }
+        }
+
+        impl Spi for $name {
+            fn init(&mut self, resets: &mut RESETS, baudrate: u32) {
+                self.spi_dev.reset_bring_down(resets);
+                self.spi_dev.reset_bring_up(resets);
+
+                self.set_baudrate(baudrate);
+                self.spi_dev
+                .sspcpsr
+                .write(|w| unsafe { w.cpsdvsr().bits(2) });
+            self.spi_dev
+                .sspcr0
+                .modify(|_, w| unsafe { w.scr().bits(3) });
+                self.set_format(8, Cpol::_0, Cpha::_0);
+                // Always enable DREQ signals -- harmless if DMA is not listening
+                self.spi_dev
+                    .sspdmacr
+                    .modify(|_, w| w.txdmae().set_bit().rxdmae().set_bit());
+
+                // Finally enable the SPI
+                self.spi_dev.sspcr1.modify(|_, w| w.sse().set_bit());
+            }
+
+            fn set_baudrate(&mut self, baudrate: u32) -> u32 {
+                let freq_in = 125_000_000; // TODO: Assumes sys clock is used and set to 125MHz
+                let mut prescale: u8 = u8::MAX;
+                let mut postdiv: u8 = 1;
+
+                // Find smallest prescale value which puts output frequency in range of
+                // post-divide. Prescale is an even number from 2 to 254 inclusive.
+                for prescale_option in (2u32..=254).step_by(2) {
+                    if freq_in < (prescale_option + 2) * 256 * baudrate {
+                        prescale = prescale_option as u8;
+                        break;
+                    }
+                }
+
+                // We might not find a prescale value that lowers the clock freq enough, so we leave it at max
+                debug_assert_ne!(prescale, u8::MAX);
+
+                // Find largest post-divide which makes output <= baudrate. Post-divide is
+                // an integer in the range 1 to 256 inclusive.
+                for postdiv_option in 256u32..1 {
```

Comment:
`256u32..1` is an empty range, you need to do something like `(1..=256u32).rev()`

Reply:
Fixed